akka-actor/src/main/scala/akka/actor/ActorSystem.scala

```diff
@@ -5,22 +5,84 @@
+   *                    threads' getContextClassLoader, then tries to walk the stack to find the callers class loader, then
+   *                    falls back to the ClassLoader associated with the ActorSystem class.
+   * @param config Configuration to use for the actor system. If no Config is given, the default reference config will be obtained from the ClassLoader.
+   * @param defaultExecutionContext If defined the ExecutionContext will be used as the default executor inside this ActorSystem.
+   *                                If no ExecutionContext is given, the system will fallback to the executor configured under "akka.actor.default-dispatcher.default-executor.fallback".
+   */
+  def apply(classLoader: Option[ClassLoader], config: Option[Config], defaultExecutionContext: Option[ExecutionContext]): ActorSystemBootstrapSettings =
+    new ActorSystemBootstrapSettings(classLoader, config, defaultExecutionContext)
+
+  /**
+   * Scala API: Short for using custom config but keeping default classloader and default execution context
+   */
+  def apply(config: Config): ActorSystemBootstrapSettings = apply(None, Some(config), None)
+
+  /**
+   * Java API: Create bootstrap settings needed for starting the actor system
+   * @param classLoader If no ClassLoader is given, it obtains the current ClassLoader by first inspecting the current
+   *                    threads' getContextClassLoader, then tries to walk the stack to find the callers class loader, then
+   *                    falls back to the ClassLoader associated with the ActorSystem class.
+   * @param config Configuration to use for the actor system. If no Config is given, the default reference config will be obtained from the ClassLoader.
+   * @param defaultExecutionContext If defined the ExecutionContext will be used as the default executor inside this ActorSystem.
+   *                                If no ExecutionContext is given, the system will fallback to the executor configured under "akka.actor.default-dispatcher.default-executor.fallback".
+   */
+  def create(classLoader: Optional[ClassLoader], config: Optional[Config], defaultExecutionContext: Optional[ExecutionContext]): ActorSystemBootstrapSettings =
+    apply(classLoader.asScala, config.asScala, defaultExecutionContext.asScala)
+
+  /**
+   * Scala API: Short for using custom config but keeping default classloader and default execution context
+   */
+  def create(config: Config): ActorSystemBootstrapSettings = apply(config)
+
+}
+
+/**
+ * Core bootstrap settings of the actor system, create using one of the factories in [[ActorSystemBootstrapSettings]].
+ *
+ * Constructor is internal API
+ */
+final class ActorSystemBootstrapSettings private[akka] (
+  val classLoader:             Option[ClassLoader],
+  val config:                  Option[Config],
+  val defaultExecutionContext: Option[ExecutionContext]) extends ActorSystemSetting {
```

Comment:
I think we want `akka.actor.provider` in here as well, makes many test nicely configurable with remoting+remote-provider settings.